core/storage/couchStorage.go

```diff
@@ -0,0 +1,1801 @@
+	ID         string    `json:"_id"`
+	Rev        string    `json:"_rev,omitempty"`
+	GroupName  string    `json:"group-name"`
+	LastUpdate time.Time `json:"last-update"`
+}
+
+type couchOrganizationObject struct {
+	ID           string              `json:"_id"`
+	Rev          string              `json:"_rev,omitempty"`
+	Organization common.Organization `json:"org"`
+	LastUpdate   time.Time           `json:"last-update"`
+}
+
+type couchACLObject struct {
+	ID         string            `json:"_id"`
+	Rev        string            `json:"_rev,omitempty"`
+	Users      []common.ACLentry `json:"users"`
+	OrgID      string            `json:"org-id"`
+	ACLType    string            `json:"acl-type"`
+	LastUpdate time.Time         `json:"last-update"`
+}
+
+// Init initializes the CouchStorage store
+func (store *CouchStorage) Init() common.SyncServiceError {
+
+	store.loginInfo = map[string]string{
+		"ipAddress": common.Configuration.CouchAddress,
+		"dbName":    common.Configuration.CouchDbName,
+		"Username":  common.Configuration.CouchUsername,
+		"Password":  common.Configuration.CouchPassword,
+	}
+
+	store.dsn = createDSN(store.loginInfo["ipAddress"])
+	var client *kivik.Client
+	var err error
+
+	client, err = kivik.New("couch", store.dsn)
+	if client == nil || err != nil {
+		message := fmt.Sprintf("Failed to connect to couch. Error: %s.", err)
+		return &Error{message}
+	}
+
+	if common.Configuration.CouchUseSSL {
+		tlsConfig := &tls.Config{}
+		if common.Configuration.CouchCACertificate != "" {
+			var caFile string
+			if strings.HasPrefix(common.Configuration.CouchCACertificate, "/") {
+				caFile = common.Configuration.CouchCACertificate
+			} else {
+				caFile = common.Configuration.PersistenceRootPath + common.Configuration.CouchCACertificate
+			}
+			serverCaCert, err := ioutil.ReadFile(caFile)
+			if err != nil {
+				if _, ok := err.(*os.PathError); ok {
+					serverCaCert = []byte(common.Configuration.CouchCACertificate)
+					err = nil
+				} else {
+					message := fmt.Sprintf("Failed to find Couch SSL CA file. Error: %s.", err)
+					return &Error{message}
+				}
+			}
+
+			caCertPool := x509.NewCertPool()
+			caCertPool.AppendCertsFromPEM(serverCaCert)
+			tlsConfig.RootCAs = caCertPool
+		}
+
+		// Please avoid using this if possible! Makes using TLS pointless
+		if common.Configuration.CouchAllowInvalidCertificates {
+			tlsConfig.InsecureSkipVerify = true
+		}
+
+		setXport := couchdb.SetTransport(&http.Transport{TLSClientConfig: tlsConfig})
+		err = client.Authenticate(context.TODO(), setXport)
+		if err != nil {
+			message := fmt.Sprintf("Authentication Failed. Error: %s.", err)
+			return &Error{message}
+		}
+	}
+
+	err = client.Authenticate(context.TODO(), &chttp.BasicAuth{Username: store.loginInfo["Username"], Password: store.loginInfo["Password"]})
+	if err != nil {
+		return err
+	}
+
+	available, err := client.Ping(context.TODO())
+	if !available || err != nil {
+		return err
```

Comment:
Is it possible that err is nill when `available == false`?

Reply:
As per my understanding, it is not possible because in the implementation of [Ping](https://github.com/go-kivik/kivik/blob/78ed46bfc99bb0c85ceca10ca3a24bf2cabb2665/kivik.go#L192), available == false is returned only when error is not nil. Please correct me if I'm wrong